### ISSUE TITLE
wss through Nginx, Error 301 no matter what I try 

### ISSUE THREAD

**Issue by manu-adina** [websockets/ws#1769]
Hi, I'm really sorry if this is the wrong place to post, I've tried StackOverflow twice, and the questions were completely unanswered. I'm not sure what the issue is when I followed the suggestions exactly.

Problem is that I keep getting Error 301 for a websocket that I'm trying to use. I've tried numerous configurations, and it seems like nothing is working for me. I'm using Amazon Lightsail, and I've enabled the port 443 and 3000, so the firewall shouldn't be stopping it (and I'm not using the AWS load balancer).

I'm trying to achieve this,
client <-- wss --> nginx <-- ws --> app server

My client side,
`const connection = new WebSocket('wss://' + location.host + '/ws');`
Server side,
`const ws = new WebSocket.Server({port: 8080});`

Nginx config file is,

```
server {
    server_name example.com;
    listen 443 ssl; # managed by Certbot                                                                                                                                                                                                         
    ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem; # managed by Certbot                                                                                                                                                                       
    ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem; # managed by Certbot                                                                                                                                             
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot                                                                                                                                                                        
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot                                                                                                                                                                  

    location /ws {
        proxy_pass http://localhost:3000;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
        proxy_http_version 1.1;                                                                                                                                                                                                        
        proxy_set_header Upgrade $http_upgrade;                                                                                                                                                                                           
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;                                                                                                                                                                                                                                                                                                                                                                                                                                           

    }

    location / {                                                                                                                                                                                                                                         
        proxy_pass http://localhost:3000;                                                                                                                                                                                                    
    }
}
server {
     if ($host = example.com) {                                                                                                                                                                                                              
          return 301 https://$host$request_uri;
     } # managed by Certbot

     listen 80;

     server_name example.com;
     return 404; # managed by Certbot
}                                                             

```

Any help would be appreciated, I've spent a long time on trying to figure this out myself but nothing is working for me :(.

**Comment by lpinca** [comment]
The only thing I can think of is this behavior

> If a location is defined by a prefix string that ends with the slash character, and requests are processed by one of proxy_pass, fastcgi_pass, uwsgi_pass, scgi_pass, memcached_pass, or grpc_pass, then the special processing is performed. In response to a request with URI equal to this string, but without the trailing slash, a permanent redirect with the code 301 will be returned to the requested URI with the slash appended.

documented in http://nginx.org/en/docs/http/ngx_http_core_module.html#location but it does not seem to reflect your configuration. Anyway try to add a leading slash to your connection URL.

**Comment by manu-adina** [comment]
Thanks for replying, I've tried the following,

`const connection = new WebSocket('wss://' + location.host + '/ws/')`

and get a websocket error of 404. I might have to scrap the whole reverse proxy, and just bind the nodejs server to port 80 which isn't recommended by people but I just can't get it working otherwise. If you think there's something else I can try to make this work, let me know, I'm eager to try it. If you're out of ideas just like me, then I'm fine with you closing this issue.

**Comment by manu-adina** [comment]
Actually, I figured it out. Turns out, for a couple of days I overlooked the fact that, my port number in the url for the proxy_pass is different from the websocket port that I set up. So a simple fix was to,

Change port number to 3001, 
`const ws = new WebSocket.Server({port: 3001});`

and have the Nginx to have the same port,
`proxy_pass http://localhost:3001;`

I've been thinking it was a problem elsewhere. Thanks for the help @lpinca , and I see you help a lot of other people too, it's really nice.
